JetBrains: ACCESS_TOKEN is null

When installing the plugin I get the following stack trace when the access token is not configured. We should report a more helpful error message instead.


```
Caused by: java.lang.IllegalArgumentException: ACCESS_TOKEN is null
	at com.sourcegraph.cody.completions.CodyCompletionsManager.completionsService(CodyCompletionsManager.java:184)
	at com.sourcegraph.cody.completions.CodyCompletionsManager.triggerCompletion(CodyCompletionsManager.java:70)
	at com.sourcegraph.cody.completions.CodyEditorFactoryListener$CodyCaretListener.caretPositionChanged(CodyEditorFactoryListener.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at com.intellij.util.EventDispatcher.dispatchVoidMethod(EventDispatcher.java:120)
	at com.intellij.util.EventDispatcher.lambda$createMulticaster$1(EventDispatcher.java:85)
	at jdk.proxy1/jdk.proxy1.$Proxy108.caretPositionChanged(Unknown Source)
	at com.intellij.openapi.editor.impl.CaretModelImpl.fireCaretPositionChanged(CaretModelImpl.java:545)
```

--- a/client/jetbrains/src/main/java/com/sourcegraph/cody/completions/CodyCompletionsManager.java
+++ b/client/jetbrains/src/main/java/com/sourcegraph/cody/completions/CodyCompletionsManager.java
@@ -17,9 +17,11 @@ import com.sourcegraph.cody.api.CompletionsService;
 import com.sourcegraph.cody.completions.prompt_library.*;
 import com.sourcegraph.cody.vscode.*;
 import com.sourcegraph.config.ConfigUtil;
+import com.sourcegraph.config.NotificationActivity;
 import com.sourcegraph.config.SettingsComponent;
 import java.util.Optional;
 import java.util.concurrent.*;
+import org.apache.commons.lang.StringUtils;
 import org.jetbrains.annotations.NotNull;
 
 /** Responsible for triggering and clearing inline code completions. */
@@ -180,8 +182,10 @@ public class CodyCompletionsManager {
     if (accessToken == null) {
       accessToken = System.getenv("SRC_ACCESS_TOKEN");
     }
-    if (accessToken == null) {
-      throw new IllegalArgumentException("ACCESS_TOKEN is null");
+    if (StringUtils.isEmpty(accessToken)) {
+      if (!ConfigUtil.isAccessTokenNotificationDismissed())
+        NotificationActivity.notifyAboutSourcegraphAccessToken(Optional.of(instanceUrl));
+      return null;
     }
     return new CompletionsService(instanceUrl, accessToken);
   }

--- a/client/jetbrains/src/main/java/com/sourcegraph/config/ConfigUtil.java
+++ b/client/jetbrains/src/main/java/com/sourcegraph/config/ConfigUtil.java
@@ -202,8 +202,16 @@ public class ConfigUtil {
     return getApplicationLevelConfig().isUrlNotificationDismissed();
   }
 
+  public static boolean isAccessTokenNotificationDismissed() {
+    return getApplicationLevelConfig().isAccessTokenNotificationDismissed();
+  }
+
   public static void setUrlNotificationDismissed(boolean value) {
     getApplicationLevelConfig().isUrlNotificationDismissed = value;
+  }
+
+  public static void setAccessTokenNotificationDismissed(boolean value) {
+    getApplicationLevelConfig().isAccessTokenNotificationDismissed = value;
   }
 
   @NotNull

--- a/client/jetbrains/src/main/java/com/sourcegraph/config/NotificationActivity.java
+++ b/client/jetbrains/src/main/java/com/sourcegraph/config/NotificationActivity.java
@@ -15,10 +15,20 @@ import com.sourcegraph.common.BrowserOpener;
 import com.sourcegraph.find.FindService;
 import java.awt.event.InputEvent;
 import java.awt.event.KeyEvent;
+import java.util.Optional;
 import javax.swing.*;
 import org.jetbrains.annotations.NotNull;
 
 public class NotificationActivity implements StartupActivity.DumbAware {
+  private static AnAction dumbAwareAction(String text, Runnable action) {
+    return new DumbAwareAction(text) {
+      @Override
+      public void actionPerformed(@NotNull AnActionEvent anActionEvent) {
+        action.run();
+      }
+    };
+  }
+
   @Override
   public void runActivity(@NotNull Project project) {
     String latestReleaseMilestoneVersion = "2.0.0";
@@ -102,5 +112,27 @@ public class NotificationActivity implements StartupActivity.DumbAware {
     Notifications.Bus.notify(notification);
 
     ConfigUtil.setLastUpdateNotificationPluginVersionToCurrent();
+  }
+
+  public static void notifyAboutSourcegraphAccessToken(Optional<String> instanceUrl) {
+    String content =
+        instanceUrl
+                .map(url -> "A Sourcegraph Access Token for " + url + " has not been set.")
+                .orElse("A Sourcegraph Access Token has not been set.")
+            + " It is not possible to use Sourcegraph API and Cody without it.";
+    Notification notification =
+        new Notification(
+            "Sourcegraph access", "Sourcegraph", content, NotificationType.INFORMATION);
+    notification.setIcon(Icons.SourcegraphLogo);
+    notification.addAction(new OpenPluginSettingsAction("Set Access Token"));
+    notification.addAction(dumbAwareAction("Do Not Set", notification::expire));
+    notification.addAction(
+        dumbAwareAction(
+            "Never Show Again",
+            () -> {
+              notification.expire();
+              ConfigUtil.setAccessTokenNotificationDismissed(true);
+            }));
+    Notifications.Bus.notify(notification);
   }
 }

--- a/client/jetbrains/src/main/java/com/sourcegraph/config/SourcegraphApplicationService.java
+++ b/client/jetbrains/src/main/java/com/sourcegraph/config/SourcegraphApplicationService.java
@@ -22,6 +22,7 @@ public class SourcegraphApplicationService
   @Nullable public String anonymousUserId;
   public boolean isInstallEventLogged;
   public boolean isUrlNotificationDismissed;
+  public boolean isAccessTokenNotificationDismissed;
   @Nullable public Boolean authenticationFailedLastTime;
 
   @Nullable
@@ -104,6 +105,10 @@ public class SourcegraphApplicationService
     return isUrlNotificationDismissed;
   }
 
+  public boolean isAccessTokenNotificationDismissed() {
+    return isAccessTokenNotificationDismissed;
+  }
+
   @Nullable
   public Boolean getAuthenticationFailedLastTime() {
     return authenticationFailedLastTime;
@@ -129,6 +134,7 @@ public class SourcegraphApplicationService
     this.remoteUrlReplacements = settings.remoteUrlReplacements;
     this.anonymousUserId = settings.anonymousUserId;
     this.isUrlNotificationDismissed = settings.isUrlNotificationDismissed;
+    this.isAccessTokenNotificationDismissed = settings.isAccessTokenNotificationDismissed;
     this.authenticationFailedLastTime = settings.authenticationFailedLastTime;
     this.lastUpdateNotificationPluginVersion = settings.lastUpdateNotificationPluginVersion;
   }

